File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/DeviceClient.swift
//
//  DeviceClient.swift
//  LevelIOSSDK
//
//  Created by Andrew Cook on 6/22/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

protocol DeviceObserverCallbacks {
    func onBluetoothNotAvailable()
    func onBluetoothNotOn()
    func onInputLedCode()
    func onLedCodeAccepted()
    func onLedCodeDone()
    func onLedCodeFailed()
    func onLedCodeNotNeeded()
    func onDeviceReady()
    func onSetUpComplete()
    func onSetUpFailed(error: ReporterError)
    func onReporterQueried(config: ReporterConfig)
    func onReportersEnabled(reporters: [ReporterType])
    func onData(data: RecordData)
    func onDataDeleted()
    func onConnectionTimeout()
    func onDisconnect()
    func onBondError()
    func onBatteryLevel(level: Int)
    func onBatteryState(state: BatteryState)
}

class DeviceClient: NSObject {
    var clientId: NSUUID
    var callbacks: DeviceObserverCallbacks?
    var bleManager: BleManager
  
    static let sharedInstance = DeviceClient()

    override init() {
        self.clientId = NSUUID()
        self.bleManager = BleManager.sharedInstance
    }

    func registerDeviceCallbacks(callbacks: DeviceObserverCallbacks) {
        self.callbacks = callbacks;
        bleManager.registerDeviceCallbacks(clientId: clientId, callbacks: callbacks);
    }

    func unregisterDeviceCallbacks() {
        bleManager.unregisterDeviceCallbacks(clientId: clientId)
    }
    
    func isConnected() -> Bool {
        return self.bleManager.isConnected()
    }

    func connect(frameId: String) {
        debugPrint("connect \(frameId) \(self.bleManager)")
        self.bleManager.connect(frameId: frameId)
    }

    func sendLedCode(code: Int) {
        debugPrint("ble manager is\(self.bleManager)")
        self.bleManager.sendLedCode(code: code)
    }
    
    func deviceLightsNotOn() {
        self.bleManager.deviceLightsNotOn()
    }
    
    func deleteKey() {
        self.bleManager.deleteSavedKey()
    }

    func getBatteryLevel() {
        self.bleManager.getBatteryLevel()
    }

    func getBatteryState() {
        self.bleManager.getBatteryState()
    }
    
    func queryReporter(reporter: ReporterType) {
        self.bleManager.queryReporter(reporter: reporter)
    }
    
    func setUpReporter(config: ReporterConfig) {
        self.bleManager.configureReporter(config: config)
    }
    
    func enableReporter(reporter: ReporterType) {
        self.bleManager.enableReporter(reporter: reporter)
    }
    
    func disableReporter(reporter: ReporterType) {
        self.bleManager.disableReporter(reporter: reporter)
    }
    
    func pauseDataStream() {
        self.bleManager.pauseData()
    }
    
    func enableDataStream() {
        self.bleManager.enableData()
    }
    
    func deleteAllStoredData() {
        self.bleManager.deleteAllData()
    }
}

File: LevelAndroidSDK/sdk/src/main/java/com/theshopatvsp/levelandroidsdk/ble/model/DeviceClient.java
package com.theshopatvsp.levelandroidsdk.ble.model;

import android.util.Log;

import com.theshopatvsp.levelandroidsdk.ble.BleManager;
import com.theshopatvsp.levelandroidsdk.ble.model.constants.BleClientCommand;
import com.theshopatvsp.levelandroidsdk.ble.model.constants.ReporterType;

import java.util.UUID;

/**
 * Created by andrco on 6/17/16.
 */
public class DeviceClient {
    private static final String TAG = DeviceClient.class.getSimpleName();
    private UUID id;
    private DeviceObserverCallbacks callbacks;

    public DeviceClient() {
        this.id = UUID.randomUUID();
    }

    /**
     * Register the device observer callbacks.
     *
     * @param callbacks - callbacks to be called when there's data from the device.
     */
    public void registerDeviceCallbacks(DeviceObserverCallbacks callbacks) {
        Log.v(TAG, "registerDeviceCallbacks called");
        this.callbacks = callbacks;

        BleManager.registerObserver(this.id, callbacks);
    }

    /**
     * Unregister callbacks to stop receiving notifications from the device.
     */
    public void unregisterDeviceCallbacks() {
        Log.v(TAG, "unregisterDeviceCallbacks called");
        BleManager.unregisterObserver(this.id);
    }

    /**
     * Connect to a specific device or connect to the closest device
     *
     * @param frameId - connect to the device ending in the frameId, set to null or the empty string to connect to the closest.
     */
    public void connect(String frameId) {
        Log.v(TAG, "connect called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.InitiateConnection, frameId));
    }

    /**
     * Disconnect from the current device.
     */
    public void disconnect() {
        Log.v(TAG, "disconnect called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.Disconnect));
    }

    /**
     * Is the device connected to the app?
     *
     * @return - true (connected) / false (not connected)
     */
    public boolean isConnected() {
        Log.v(TAG, "isConnected called");
        return BleManager.isConnected();
    }

    /**
     * Send the led code from Blink to link to the device
     *
     * @param code - Hex code of the led color
     */
    public void sendLedCode(int code) {
        Log.v(TAG, "sendLedCode called: " + code);
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.SendLedCode, code));
    }

    /**
     * If the led on the device has not light up, call this method to go to the next closest device
     */
    public void deviceLightsNotOn() {
        Log.v(TAG, "deviceLightsNotOn called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.DeviceLightsNotOn));
    }

    /**
     * Get the battery level from the device
     */
    public void getBatteryLevel() {
        Log.v(TAG, "getBatteryLevel called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.GetBatteryLevel));
    }

    /**
     * Get the battery state of the device, @see com.theshopatvsp.levelandroidsdk.ble.model.constants.BatteryState
     */
    public void getBatteryState() {
        Log.v(TAG, "getBatteryState called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.GetBatteryState));
    }

    /**
     * Query the configuration of the given reporter type
     *
     * @param type - the reporter type to query
     */
    public void queryReporter(ReporterType type) {
        Log.v(TAG, "queryReporter called: " + type);
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.QueryReporter, type.getReporter()));
    }

    /**
     * Configure the device reporters to specific settings
     *
     * @param config - config object @see com.theshopatvsp.levelandroidsdk.ble.model.ReporterConfig
     */
    public void setUpDevice(DeviceConfig config) {
        Log.v(TAG, "setUpDevice called: " + config.toString());
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.SetUpReporter, config));
    }

    /**
     * Turn off the reporter to keep it from generating data
     *
     * @param type - the reporter type to turn off
     */
    public void disableReporter(ReporterType type) {
        Log.v(TAG, "disableReporter called: " + type);
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.DisableReporter, type));
    }

    /**
     * Turn on the reporter to get it to start generating data
     *
     * @param type - the reporter type to turn off
     */
    public void enableReporter(ReporterType type) {
        Log.v(TAG, "enableReporter called: " + type);
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.EnableReporter, type));
    }

    public void queryEnabledReporters() {
        Log.v(TAG, "queryEnabledReporters called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.QueryReportControl));
    }

    /**
     * Turn on the data stream to download the generated data from the device
     */
    public void enableDataStream() {
        Log.v(TAG, "enableDataStream called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.EnableDataStream));
    }

    /**
     * Turn off the data stream to download the generated data from the device
     */
    public void disableDataStream() {
        Log.v(TAG, "disableDataStream called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.DisableDataStream));
    }

    /**
     * Delete all fo the stored data from the device
     */
    public void deleteAllStoredData() {
        Log.v(TAG, "deleteAllStoredData called");
        BleManager.addClientCommand(new ClientCommand(BleClientCommand.DeleteStoredData));
    }
}
